[CORE-13215] ct/l1: Add a level one reader

wdberkeley opened the pull request:
Basically follows the design of the L0 reader, which works well and is easy to understand.

## Backports Required

- [X] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v25.2.x
- [ ] v25.1.x
- [ ] v24.3.x

## Release Notes

* none

wdberkeley committed:
ct/l1: Add l1 reader

This adds an L1 reader. Fundamentally it follows the same flow as the L0
reader. More extensive documentation can be found in the header.

WillemKauf commented on src/v/cloud_topics/level_one/frontend_reader/reader.cc:
```diff
@@ -0,0 +1,439 @@
+  model::timeout_clock::time_point deadline) {
+    if (cache_enabled() && _state == state::empty) {
+        if (auto cached = maybe_load_slices_from_cache()) {
+            co_return std::move(cached.value());
```
nit: `co_return std::move(cached).value();` though it probably doesn't make a difference here.

rockwotj commented on src/v/cloud_topics/level_one/frontend_reader/reader.h:
```diff
@@ -0,0 +1,122 @@
+    std::optional<l1::object_id> _oid;
+
+    // Cached footer for the current object.
+    std::optional<l1::footer> _footer;
```
should these be a single `std::optional` that is a struct containing both of these things?

rockwotj replied:
Ping on this

dotnwat replied:
> should these be a single std::optional that is a struct containing both of these things?

something like this?

```c++
struct object { oid, footer };
std::optional<object> current;
```

rockwotj commented on src/v/cloud_topics/level_one/frontend_reader/reader.cc:
```diff
@@ -0,0 +1,439 @@
+        size_t batch_size = batch.value().size_bytes();
+
+        if (
+          !ret.empty() && batch_size + materialized_bytes > _config.max_bytes) {
```
there is a strict limit flag were we return no results if that flag is set

rockwotj commented on src/v/cloud_topics/level_one/frontend_reader/reader.cc:
```diff
@@ -0,0 +1,439 @@
+      .size = footer_total_size,
+    };
+
+    auto stream_result = co_await _io->read_object(extent, nullptr);
```
passing in `nullptr` here is UB because it's rereferenced

rockwotj commented on src/v/cloud_topics/level_one/frontend_reader/reader.cc:
```diff
@@ -0,0 +1,439 @@
+
+    auto& stream = stream_result.value();
+    iobuf footer_buf = co_await read_iobuf_exactly(stream, footer_total_size);
+    co_await stream.close();
```
The stream close method must be called even if there is an exception.

WillemKauf replied:
+1, `ss::coroutine::as_future()` here would be ideal.

Part of me wonders if there should be an owning `read_iobuf_exactly_and_close_stream(ss::input_stream<>&&, size_t)` to avoid the manual toil, but I looked through the codebase and mostly found cases where the lifetime of the stream is longer than the case here.

rockwotj replied:
We should probably add another method to L1 IO for reading into an iobuf (instead of a stream). I have a TODO for it somewhere...

rockwotj commented on src/v/cloud_topics/level_one/frontend_reader/reader.cc:
```diff
@@ -0,0 +1,439 @@
+        vlog(
+          cd_log.error,
+          "Failed to parse footer from object {} despite reading complete "
+          "footer",
```
nit add position and file size to this log message?

rockwotj commented on src/v/cloud_topics/level_one/frontend_reader/reader.cc:
```diff
@@ -0,0 +1,439 @@
+              _ntp,
+              _tidp,
+              _oid);
+            _state = state::end_of_stream;
```
should we be throwing instead?

rockwotj replied:
I think this is currently how the read path propagates errors, so throwing is better then silently truncating the partition.

rockwotj commented on src/v/cloud_topics/level_one/frontend_reader/reader.cc:
```diff
@@ -0,0 +1,439 @@
+        l1::object_extent extent{
+          .id = _oid.value(),
+          .position = start_pos,
+          .size = L1_max_bytes_per_object_fetch,
```
This value is much too small for data, it's OK in L0 because we're fetching metadata.

rockwotj commented on src/v/cloud_topics/level_one/frontend_reader/reader.cc:
```diff
@@ -0,0 +1,439 @@
+                break;
+            } else if (
+              std::holds_alternative<l1::footer>(result)
+              || std::holds_alternative<l1::object_reader::eof>(result)) {
```
nit: `ss::visit` instead of the if elses? you can return `ss::stop_iteration` to know if you should keep going.

rockwotj commented on src/v/cloud_topics/level_one/frontend_reader/reader.cc:
```diff
@@ -0,0 +1,439 @@
+                size_t batch_size = batch.size_bytes();
+                if (
+                  _config.bytes_consumed > 0
+                  && _config.bytes_consumed + batch_size > _config.max_bytes) {
```
need to handle the strict flag

rockwotj commented on src/v/cloud_topics/level_one/frontend_reader/reader.cc:
```diff
@@ -0,0 +1,439 @@
+                // Cache if needed.
+                if (cache_enabled()) {
+                    _ct_api->cache_put(_ntp, batch.copy());
+                }
```
we can cache even if we throw the batch away

rockwotj commented on src/v/cloud_topics/level_one/frontend_reader/reader.cc:
```diff
@@ -0,0 +1,439 @@
+                if (
+                  batch.last_offset()
+                  < kafka::offset_cast(_config.start_offset)) {
+                    continue;
```
There is a lot of nesting going on which means you have to keep a lot of context in mind as you read this function. Can we break this all up into the logical pieces so it's easier to read? Maybe this means we move the `try/catch` up a level or something.

wdberkeley replied:
This changed a bit with the exception handling change and the missing offsets change. Let me know if you think it still needs some refactoring 👍

rockwotj commented on src/v/cloud_topics/level_one/frontend_reader/reader.cc:
```diff
@@ -0,0 +1,439 @@
+            } else if (std::holds_alternative<model::ntp>(result)) {
+                // Partition marker. Done with this ntp's partition.
+                // TODO: The next marker could be the same ntp.
+                //       Could micro-optimize for this.
```
no because there is nothing saying these NTPs are in order

rockwotj commented on src/v/cloud_topics/level_one/frontend_reader/tests/reader_test.cc:
```diff
@@ -0,0 +1,597 @@
+
+// Some of this is borrowed from log segment reader tests.
+template<typename l_iter_t, typename r_iter_t>
+void check_iters(
+  l_iter_t l_iter, l_iter_t l_end, r_iter_t r_iter, r_iter_t r_end) {
+    bool l_done = false;
+    bool r_done = false;
+    while (true) {
+        l_done = l_iter == l_end;
+        r_done = r_iter == r_end;
+        if (l_done || r_done) {
+            break;
+        }
+        EXPECT_EQ(*r_iter, *l_iter);
+        ++r_iter;
+        ++l_iter;
+    }
+    EXPECT_EQ(l_done, r_done);
+}
```
could be simplified with `std::ranges::zip`

WillemKauf replied:
`std::deque` (our typedef for `chunked_circular_buffer`) defines `operator==`, so is this `check_batches()`/`check_iters()` even necessary instead of just `EXPECT_EQ(actual, expected)`?

rockwotj reviewed (commented):
Need to go over the tests in depth, but this is a good first round

Lazin commented on src/v/cloud_topics/level_one/frontend_reader/reader.cc:
```diff
@@ -0,0 +1,439 @@
+level_one_log_reader_impl::level_one_log_reader_impl(
+  cloud_topic_log_reader_config& cfg,
+  model::ntp ntp,
+  model::topic_id_partition tidp,
```
why do we need both ntp and tidp here?

rockwotj replied:
I think we need to update the l1 object builder/reader to use topic ids instead of ntps. I was going to do that and forgot

rockwotj replied:
https://github.com/redpanda-data/redpanda/pull/27453

wdberkeley replied:
Thanks!

Lazin commented on src/v/cloud_topics/level_one/frontend_reader/reader.cc:
```diff
@@ -0,0 +1,439 @@
+      .size = footer_total_size,
+    };
+
+    auto stream_result = co_await _io->read_object(extent, nullptr);
```
are we reading from the cloud storage cache here or directly from the cloud?

rockwotj replied:
the IO object handles both caching and the read from object storage

andrwng commented on src/v/cloud_topics/level_one/frontend_reader/reader.h:
```diff
@@ -0,0 +1,122 @@
+    model::topic_id_partition _tidp;
+    l1::metastore* _metastore;
+    l1::io* _io;
+    data_plane_api* _ct_api;
```
For non-compacted topics this is probably fine, but I think it generally won't do the right thing for compacted topics -- thinking about cases where tombstones are removed and the batch cache doesn't have a way to remove old records that should have been deduplicated.

I'd suggest for now just leaving the batch cache out of the equation -- maybe we can add a different batch cache in a different layer that's keyed by object ID or something to be able to get more compaction-consistent semantics when reading

andrwng commented on src/v/cloud_topics/level_one/frontend_reader/reader.cc:
```diff
@@ -0,0 +1,439 @@
+    switch (_state) {
+    case state::empty:
+    case state::ready:
+        _state = state::end_of_stream;
```
We chatted about this a little offline, I think it'll be important to distinguish between "end of stream" caused by an error and "end of stream" where we've read everything the metastore told us to. Especially for compacted topics, where reading everything may not necessarily net a reader that covers all offsets, it feels like it'll be useful to distinguish if we're missing offsets because of compaction or because there was some error

andrwng replied:
Another case to consider: detecting out_of_range errors specifically, as that may indicate that we've successfully read everything from L1, and the next caller should consumer from L0.

andrwng commented on src/v/cloud_topics/level_one/frontend_reader/reader.cc:
```diff
@@ -0,0 +1,439 @@
+        co_return;
+    }
+
+    _state = _batches.empty() ? state::end_of_stream : state::materialized;
```
Curious if this should be empty instead of end_of_stream? I'm thinking about the case of reading a compacted range of data that has no records between the given offsets. Does this end up making progress? Do we need to refetch metadata at the next offset past the current extent or something?

WillemKauf commented on src/v/cloud_topics/level_one/frontend_reader/reader.cc:
```diff
@@ -0,0 +1,439 @@
+          .position = start_pos,
+          .size = L1_max_bytes_per_object_fetch,
+        };
+        auto stream_result = co_await _io->read_object(extent, nullptr);
```
Also needs a non-null `abort_source*`. See: 

https://github.com/redpanda-data/redpanda/blob/920219888876bf256c983aecc0a7e1945d89c3f6/src/v/cloud_topics/level_one/common/file_io.cc#L151-L154

WillemKauf commented on src/v/cloud_topics/level_one/frontend_reader/reader.cc:
```diff
@@ -0,0 +1,439 @@
+        }
+
+        auto reader = l1::object_reader::create(
+          std::move(stream_result.value()));
```
nit: `std::move(stream_result).value()`

WillemKauf commented on src/v/cloud_topics/level_one/frontend_reader/reader.cc:
```diff
@@ -0,0 +1,439 @@
+            }
+        }
+
+        co_await reader->close();
```
Also need some exception safety on above block of code and this `close()` call I think

WillemKauf commented on src/v/cloud_topics/level_one/frontend_reader/reader.cc:
```diff
@@ -0,0 +1,439 @@
+
+        vassert(
+          batch->base_offset() == kafka::offset_cast(current),
+          "Unexpected base offset {} vs {}",
```
what if compacted topic?

WillemKauf commented on src/v/cloud_topics/level_one/frontend_reader/reader.cc:
```diff
@@ -0,0 +1,439 @@
+          batch->base_offset(),
+          current);
+
+        ret.push_back(std::move(batch.value()));
```
nit: `std::move(batch).value()`

rockwotj commented on src/v/cloud_topics/log_reader_config.h:
```diff
@@ -87,6 +87,8 @@ struct cloud_topic_log_reader_config {
     // that the reader will return no batches.
     bool strict_max_bytes{false};
 
+    bool over_budget{false};
```
do we need to add this? Shouldn't we be able to use a predicate like: `bytes_consumed >= max_bytes`

rockwotj commented on src/v/cloud_topics/level_one/frontend_reader/reader.cc:
```diff
@@ -0,0 +1,415 @@
+    auto footer_result = co_await read_footer(
+      obj.oid, obj.footer_pos, obj.object_size);
+    if (!footer_result.has_value()) {
+        vlog(
+          cd_log.warn,
+          "Failed to read footer from L1 object {} for {} ({}): {}",
+          obj.oid,
+          _ntp,
+          _tidp,
+          std::to_underlying(footer_result.error()));
+        _state = state::end_of_stream;
+        co_return;
```
I do not think we want to silently truncate the stream when there are errors, but instead throw or return some error so the client gets an error

rockwotj commented on src/v/cloud_topics/level_one/frontend_reader/reader.cc:
```diff
@@ -0,0 +1,415 @@
+      _batches.empty(),
+      "Materialize batches called with batches already materialized");
+
+    if (_state == state::end_of_stream) {
+        co_return;
+    }
```
shouldn't this only be called if `_state == state::ready`?

rockwotj replied:
ping - should this be a vassert?

wdberkeley replied:
The top-level double-switch logic always calls `materialize_batches` after `fetch_metadata`, but `fetch_metadata` can transition to end-of-stream.

dotnwat replied:
>The top-level double-switch logic always calls materialize_batches after fetch_metadata, but fetch_metadata can transition to end-of-stream.

it might be simpler to understand if all the state transition logic (like this guard) were centralized into the switch statement, and only the pre-conditions were checked on entry to these routines.

but it's probably not worth changes. this state machine isn't too big for it matter.

rockwotj commented on src/v/cloud_topics/level_one/frontend_reader/reader.cc:
```diff
@@ -0,0 +1,415 @@
+      "Invalid state to materialize batches: {}",
+      std::to_underlying(_state));
+
+    // I wish I had ADTs to enforce these invariants...
```
We don't need to do this we probably should get this in, but you could use a variant, and `ss::visit` instead of the switch statements.

```
struct empty{};
struct current_object {/*footer + oid*/};
struct materialized {optional<current_object> if there is more to read>, batches};
struct eos{};
using state = std::variant<empty, current_object, materialized, eos>
```

rockwotj commented on src/v/cloud_topics/level_one/frontend_reader/reader.cc:
```diff
@@ -0,0 +1,415 @@
+          _ntp,
+          _tidp,
+          _current_obj->oid);
+        _state = state::materialized;
```
Don't we need to advance offsets in this case? Otherwise we'll just loop endlessly.

wdberkeley replied:
`consume_materialized_batches` will always advance the offset.

dotnwat replied:
makes sense

rockwotj commented on src/v/cloud_topics/level_one/frontend_reader/reader.cc:
```diff
@@ -0,0 +1,415 @@
+namespace cloud_topics {
+
+// TODO: Add config.
+static constexpr size_t L1_max_bytes_per_object_fetch = 1024_KiB;
```
This should be 4MiB at least. I think current TS read path uses 16MiB maybe?

rockwotj commented on src/v/cloud_topics/level_one/frontend_reader/reader.cc:
```diff
@@ -0,0 +1,415 @@
+    l1::object_extent extent{
+      .id = _current_obj->oid,
+      .position = start_pos,
+      .size = L1_max_bytes_per_object_fetch,
```
Thinking more, we'll probably think we get corrupted data in the L1 reader if we're reading a partial object because it was truncated in the middle. I think we *have* to pick boundaries that are defined in the footer, or we need to make modifications to the l1 object reader to understand it could be called with partial data.

My heavy preference is to just always align on boundaries for now, supporting partial data in the reader will complicate the code a bunch to distinguish this case between real corruption.

How do the tests not catch this? Are all the batches super small?

wdberkeley replied:
Yeah, I noticed this as well. We don't run into it now because the partitions are smaller than the fetch size. To handle this I'd think we'd want the footer's API to return not just the position from `file_position_before_{kafka_offset,max_timestamp`, but also the size of the rest of the partition.

rockwotj replied:
https://github.com/redpanda-data/redpanda/pull/27554

I think we'll want a different interface for better caching, but this is a step in the right direction

rockwotj commented on src/v/cloud_topics/level_one/frontend_reader/reader.cc:
```diff
@@ -0,0 +1,415 @@
+                _config.over_budget = true;
+                break;
+            }
+            _config.bytes_consumed += batch_size;
```
BTW I find it strange that bytes are incremented here, but offsets are incremented in another place.

wdberkeley replied:
The offset accounting happens automatically by advancing through batches. The other place updates the config's start offset to move on to the next "batch of batches".

rockwotj commented on src/v/cloud_topics/level_one/frontend_reader/reader.cc:
```diff
@@ -0,0 +1,415 @@
+      _ntp,
+      _tidp);
+
+    std::move(_batches.begin(), _batches.end(), std::back_inserter(*dest));
+    _batches.clear();
```
why not keep _batches in the same data structure as `dest` so they can just be swapped out instead of just moving the container?

rockwotj commented on src/v/cloud_topics/level_one/frontend_reader/reader.cc:
```diff
@@ -0,0 +1,415 @@
+  l1::metastore* metastore,
+  l1::io* io_interface)
+  : _config(cfg)
+  , _ntp(std::move(ntp))
```
since NTP is just used for logging do we really need it?

wdberkeley replied:
No, I just thought it'd be nice to have the name in the logs rather than a UUID.

dotnwat replied:
>No, I just thought it'd be nice to have the name in the logs rather than a UUID.

yeh good point. we'll need to revisit our logging across all of cloud topics too and probably think about some of these details.

rockwotj commented on src/v/cloud_topics/level_one/frontend_reader/tests/reader_test.cc:
```diff
@@ -0,0 +1,571 @@
+
+    for (auto start = 0; start <= 200; start += 50) {
+        std::vector<tidp_batches_t> tidp_batches;
+        for (auto i = 0; i < 3; i++) {
```
std::ranges::zip is an option too

wdberkeley replied:
Good point. It's test code that's pretty easy to understand and already works so I'll leave it be unless you feel strongly about zipping.

rockwotj commented on src/v/cloud_topics/level_one/frontend_reader/tests/reader_test.cc:
```diff
@@ -0,0 +1,571 @@
+
+    auto reader = make_reader(ntp, tidp);
+    auto result = read_all(reader.get());
+    EXPECT_TRUE(result.empty());
+    EXPECT_TRUE(reader->is_end_of_stream());
```
This feels very misleading and I think we should throw

rockwotj commented on src/v/cloud_topics/level_one/frontend_reader/tests/reader_test.cc:
```diff
@@ -0,0 +1,571 @@
+
+    _io.put_object(oid, std::move(buf));
+
+    auto missing = kafka::next_offset(
+      model::offset_cast(batches.back().last_offset()));
```
nit: `missing` is a bit confusing to me because offset -1 also means "missing" in the kafka API. Can we call this high_watermark or something a bit more standard in kafka terms?

rockwotj commented on src/v/cloud_topics/level_one/frontend_reader/tests/reader_test.cc:
```diff
@@ -0,0 +1,571 @@
+          .base_offset = missing,
+          .last_offset = missing,
+          .max_timestamp = model::timestamp::now(),
+          .pos = 0,
+          .size = 0,
```
is this the intended values for an empty range @andrwng?

wdberkeley replied:
FWIW this is based on nothing from me... I just put zeroes and hoped it worked. Seems like we do need rules for empty ranges in case compaction wipes out a whole range.

andrwng replied:
Yeah this is my expectation, that the size is 0 for an empty range. And I'd expect the object footer to not include the tidp at all.

rockwotj commented on src/v/cloud_topics/level_one/frontend_reader/tests/reader_test.cc:
```diff
@@ -0,0 +1,571 @@
+    EXPECT_TRUE(reader->is_end_of_stream());
+}
+
+TEST_F(l1_reader_test, sparse_offset_ranges) {
```
Not needed right now, but this kind of thing is also useful to have randomized testing that creates a random amount of data with randomized boundaries and at various different offset boundaries matches what the reader returns.

dotnwat commented on src/v/cloud_topics/level_one/frontend_reader/reader.h:
```diff
@@ -0,0 +1,122 @@
+This class implements a record batch reader for level one.
+
+The reader is a state machine with the following states:
+- empty: no metadata is cached, no data is materialized
```
Can you reflow this comment so each lines contains a `*` prefix?

dotnwat commented on src/v/cloud_topics/level_one/frontend_reader/reader.h:
```diff
@@ -0,0 +1,122 @@
+
+In 'materialized' state, data can be consumed. When all materialized
+batches are consumed, the reader transitions back to 'empty' or 'ready'
+state depending on whether more data is available in the current object.
```
It's probably out of scope for this PR, but this state machine seems like it doesn't leave much room for read-ahead?

wdberkeley replied:
The metastore API doesn't allow for read-ahead. The API returns information about the object containing the lowest offset >= a given offset, so we need to know the offset we are ending at to query for the next object (or tip-toe through offset space looking for the next object). Andrew and I have talked about another API that would return information about a range of offsets, so the reader could get information about the next <= N objects that cover the offset range [start, end] or [start, infinity), and then it'd be possible to read-ahead.

dotnwat commented on src/v/cloud_topics/level_one/frontend_reader/reader.h:
```diff
@@ -0,0 +1,122 @@
+      l1::metastore* metastore,
+      l1::io* io_interface,
+      // TODO: This is just for the cache. Maybe pass that directly?
+      data_plane_api* ct_api);
```
unused?

dotnwat commented on src/v/cloud_topics/level_one/frontend_reader/reader.cc:
```diff
@@ -0,0 +1,439 @@
+  , _ct_api(ct_api) {
+    if (_config.max_bytes == 0) {
+        _state = state::end_of_stream;
+    }
```
* maybe remove this and handle it in a unified code path, just to reduce complexity...?
* the other thing to consider is if max bytes applies before or after the first batch. for example, does max bytes = 0 mean we read no batches, or no more than one? i'd use whatever the semantics are in disk_log_impl reader.

wdberkeley replied:
Yeah this is a hand-me-down from the level_zero reader code I based the level_one reader on but this isn't right when max_bytes isn't strict.

dotnwat commented on src/v/cloud_topics/level_one/frontend_reader/reader.h:
```diff
@@ -0,0 +1,122 @@
+    void print(std::ostream& o) final;
+
+private:
+    enum class state {
+        empty,
+        ready,
+        materialized,
+        end_of_stream,
+    };
```
i am little curious how much you think the states simplify the implementation at this point. for example, it seems like empty/read/materialized effectively map 1-1 onto other state tracked by the reader (e.g. is the slice empty, do i have a cached "current object"), etc...

wdberkeley replied:
True it's mostly vestigial from the level zero reader this is based on, but I still like it conceptually because it divides up the work of the reader into phases.

dotnwat commented on src/v/cloud_topics/level_one/frontend_reader/reader.cc:
```diff
@@ -0,0 +1,439 @@
+level_one_log_reader_impl::do_load_slice(
+  model::timeout_clock::time_point deadline) {
+    if (cache_enabled() && _state == state::empty) {
+        if (auto cached = maybe_load_slices_from_cache()) {
```
please use the C++17 init-statement form for the conditional

dotnwat commented on src/v/cloud_topics/level_one/frontend_reader/reader.cc:
```diff
@@ -0,0 +1,439 @@
+
+    chunked_circular_buffer<model::record_batch> res;
+    switch (_state) {
+    case state::empty:
+        co_await fetch_metadata(deadline);
```
if the check for cached data were inserted here, would it eliminate the special case above?

wdberkeley replied:
Do you mean the max_bytes check?

dotnwat replied:
this is an example it looks like where I reviewed an older version of this file in the commit history.

dotnwat commented on src/v/cloud_topics/level_one/frontend_reader/reader.cc:
```diff
@@ -0,0 +1,439 @@
+ss::future<model::record_batch_reader::storage_t>
+level_one_log_reader_impl::do_load_slice(
+  model::timeout_clock::time_point deadline) {
+    if (cache_enabled() && _state == state::empty) {
```
i would have assumed that caching is active no matter what state we are in, but if anything, it would seem to be _less_ likely to be active in the empty state given that the empty state is where we acquire metadata (which would tell us what the identify of data in the cache are)?

it would be helpful to address this in the comment on the reader class.

dotnwat commented on src/v/cloud_topics/level_one/frontend_reader/reader.cc:
```diff
@@ -0,0 +1,439 @@
+
+    // Invariant: in case of success the state will be either materialized
+    // or end_of_stream. In case of error the state will be end_of_stream.
+    switch (_state) {
```
i'm a bit confused why there are two switch statements. could you add comment explaining that? it seems like either they could be merged, or put into a loop, or rely on the caller to re-invoke load slice?

wdberkeley replied:
I can add a comment. Basically it handles moving the reader from wherever it is into a materialized state without looping, enforcing that we end up EOS or materialized after fetching and materializing.

dotnwat commented on src/v/cloud_topics/level_one/frontend_reader/reader.cc:
```diff
@@ -0,0 +1,439 @@
+    vassert(
+      !_footer.has_value(), "Empty state should not have a current footer");
+
+    if (_config.bytes_consumed >= _config.max_bytes) {
```
i think this will handle the case in the constructor

dotnwat commented on src/v/cloud_topics/log_reader_config.h:
```diff
@@ -87,6 +87,8 @@ struct cloud_topic_log_reader_config {
     // that the reader will return no batches.
     bool strict_max_bytes{false};
 
+    bool over_budget{false};
```
from the name alone, it doesn't seem like this belongs in the reader config--like would i want to create a reader that was immediately over budget? seems like maybe if you are using it to store state across some contexts, that that should be solved separately.

wdberkeley replied:
I went this way because it seems to be a standard part of reader configs. The cloud storage and log_reader both have `over_budget` in their configs and use it for the same purpose.

dotnwat commented on src/v/cloud_topics/level_one/frontend_reader/reader.cc:
```diff
@@ -0,0 +1,415 @@
+          _current_obj->oid,
+          _ntp,
+          _tidp,
+          stream_fut.get_exception());
```
we should avoid assuming that vlog always evaluates its parameters (e.g. some day it might be evolved to a macro that checks log level first) and extract the exception separately. kind of like avoiding putting side effects in an assert.

dotnwat commented on src/v/cloud_topics/level_one/frontend_reader/reader.cc:
```diff
@@ -0,0 +1,415 @@
+    co_return std::get<l1::footer>(std::move(footer_result));
+}
+
+ss::future<> level_one_log_reader_impl::materialize_batches(
```
there are a lot of comments in this function, so i can't tell exactly how long it is, but it seems pretty long. i wonder if if it makes sense to split into two where the top half gets the reader and has a try/catch so that it is reliably closed and then the inner part does the rest?

wdberkeley replied:
Good idea. It decomposed naturally into the setup part and the batch reading part 👍 .

dotnwat commented on src/v/cloud_topics/level_one/frontend_reader/reader.cc:
```diff
@@ -0,0 +1,415 @@
+          oid,
+          extent.position,
+          object_size,
+          stream_fut.get_exception());
```
same here. avoid side effects in vlog parameters.

dotnwat commented on src/v/cloud_topics/level_one/frontend_reader/reader.cc:
```diff
@@ -0,0 +1,415 @@
+        co_return;
+    }
+
+    if (_config.over_budget) {
```
can over_budget be reader state?

wdberkeley replied:
See other comment. It could be, but we have a pattern of putting it in configs.

dotnwat reviewed (commented):
sorry some of my comments may be out-dated because i thought changes you made in later commits would have been squashed back down. my bad

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--df1f536e-59e4-4966-9236-208eee2cd1cb-->
<!--TEST RESULTS-->
<!--5456220d-33d1-41d8-ba72-f266bde53f60-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/72025">build#72025</a></summary>


| test_class                        | test_method                 | test_arguments             | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                    |
|:----------------------------------|:----------------------------|:---------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:----------------------------------------------------------------------------------------------------------------------------------------------------------|
| VerifyConsumerOffsetsThruUpgrades | test_consumer_group_offsets | {"versions_to_upgrade": 3} | integration | https://buildkite.com/redpanda/redpanda/builds/72025#01993526-a59d-4a12-802e-a74e7c1d5ac7 | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/72277">build#72277</a></summary>


| test_class               | test_method                         | test_arguments                                                                                                                       | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                  |
|:-------------------------|:------------------------------------|:-------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| CloudStorageScrubberTest | test_scrubber                       | {"cloud_storage_type": 1}                                                                                                            | integration | https://buildkite.com/redpanda/redpanda/builds/72277#01994e80-1db8-48ee-b93d-0dae600292fe | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS               |
| DatalakeE2ETests         | test_json_schema_unicode            | {"catalog_type": "rest_hadoop", "cloud_storage_type": 1, "query_engine": "trino"}                                                    | integration | https://buildkite.com/redpanda/redpanda/builds/72277#01994e79-3751-4575-bf0c-a9647be21b18 | FLAKY         | 20/21    | upstream reliability is '99.64912280701755'. current run reliability is '95.23809523809523'. drift is 4.41103 and the allowed drift is set to 50. The test should PASS  |
| SchemaEvolutionE2ETests  | test_old_schema_writer              | {"catalog_type": "rest_hadoop", "cloud_storage_type": 1, "produce_mode": "avro", "query_engine": "trino", "test_case": "add_column"} | integration | https://buildkite.com/redpanda/redpanda/builds/72277#01994e80-1db3-4e28-b5a6-14b1770f9e52 | FLAKY         | 20/21    | upstream reliability is '99.3421052631579'. current run reliability is '95.23809523809523'. drift is 4.10401 and the allowed drift is set to 50. The test should PASS   |
| PartitionBalancerTest    | test_recovery_mode_rebalance_finish | null                                                                                                                                 | integration | https://buildkite.com/redpanda/redpanda/builds/72277#01994e79-3751-4ef9-b70e-306bf8d2081c | FLAKY         | 15/21    | upstream reliability is '87.00361010830325'. current run reliability is '71.42857142857143'. drift is 15.57504 and the allowed drift is set to 50. The test should PASS |
</details>
<!--/body-->

andrwng commented on src/v/cloud_topics/level_one/frontend_reader/reader.cc:
```diff
@@ -0,0 +1,443 @@
+            object_size));
+    }
+
+    co_return std::get<l1::footer>(std::move(footer_result));
```
Optimization thought (not for this PR): maybe we should shrink down the footer to only include the partitions relevant to the tidp

andrwng commented on src/v/cloud_topics/level_one/frontend_reader/reader.cc:
```diff
@@ -0,0 +1,443 @@
+
+    _config.start_offset
+      = dest->empty()
+          ? kafka::next_offset(_config.start_offset)
```
Optimization thought (not for this PR): it shouldn't be too hard to have the metastore return the last offset for the fetch extent. Then here we could jump to `kafka::next_offset(_current_obj.last_offset)`

andrwng commented on src/v/cloud_topics/level_one/frontend_reader/reader.cc:
```diff
@@ -0,0 +1,443 @@
+    }
+
+    auto response_fut = co_await ss::coroutine::as_future(
+      _metastore->get_first_ge(_tidp, _config.start_offset));
```
Optimization thought (not for this PR): would be cool to extend the metastore interface to return the next N extents (up to the max offset) so we don't have to keep going back to the metastore

andrwng commented on src/v/cloud_topics/level_one/frontend_reader/reader.cc:
```diff
@@ -0,0 +1,443 @@
+    _state = state::ready;
+}
+
+ss::future<l1::footer> level_one_log_reader_impl::read_footer(
```
nit (maybe in a follow up though): I think this would be better placed in file_io, because then we'd be able to transparently cache the footers in a way that is reusable by multiple readers quite easily

rockwotj replied:
yes agreed, there is a todo in file IO for it

andrwng commented on src/v/cloud_topics/level_one/frontend_reader/reader.cc:
```diff
@@ -0,0 +1,443 @@
+
+    auto response = response_fut.get();
+    if (!response.has_value()) {
+        if (response.error() == l1::metastore::errc::out_of_range) {
```
nit: use a switch?

andrwng reviewed (commented):
Overall this lgtm. Left a few comments about improvements, but none are blocking.

dotnwat commented on src/v/cloud_topics/level_one/frontend_reader/reader.h:
```diff
@@ -0,0 +1,122 @@
+ * In 'materialized' state, data can be consumed. When all materialized
+ * batches are consumed, the reader transitions back to 'empty' state.
+ * The reader queries the metastore after each L1 object partition that
+ * it processes.
```
is this describing what happens in the empty state, or the materialized state?

wdberkeley replied:
Back in empty state. I update the wording to reflect this.

dotnwat commented on src/v/cloud_topics/level_one/frontend_reader/tests/reader_test.cc:
```diff
@@ -0,0 +1,620 @@
+    return result;
+}
+
+// Slice by offset range (copy-based, preserves original)
```
just to be clarify, the intention here is copy whole batches that intersect with the offset range, rather than the more complicated form of this which would be to split batches?

wdberkeley replied:
Yep. Comment updated.

dotnwat commented on src/v/cloud_topics/level_one/frontend_reader/tests/reader_test.cc:
```diff
@@ -0,0 +1,620 @@
+              batches.begin(), batches.end(), std::back_inserter(all_batches));
+        }
+        return all_batches;
+    }
```
I don't know if this was already mentioned before in reviews, but interacting with the `::impl` directly is not the way a record batch reader is typically used. Perhaps there is some value in testing the `::impl` directly, but I think we should have at least some cases that cover the common form (if not all cases converted over).

The transformation I think is straight forward. For `make_reader`:

```c++
model::record_batch_reader make_reader() {
  return model::record_batch_reader(std::make_unique<level_one_log_reader_impl>(
          config, ntp, tidp, &_metastore, &_io))
}
```

And for `read_all`:

```c++
chunked_circular_buffer<model::record_batch>
read_all(model::record_batch_reader reader) {
   auto data = co_await model::consume_reader_to_memory(std::move(reader), model::no_timeout);
   // TODO: convert data into circular buffer
}
```

dotnwat commented on src/v/cloud_topics/level_one/frontend_reader/reader.cc:
```diff
@@ -0,0 +1,443 @@
+            co_return;
+        }
+        if (response.error() == l1::metastore::errc::missing_ntp) {
+            vlog(
+              cd_log.debug,
+              "Partition {} ({}) not tracked in metastore",
+              _ntp,
```
nit: you can use the `prefix logger` utility (no reason to do that in this PR tho) that we have so that NTP is prefixed automatically and you don't have to manually add it.

dotnwat commented on src/v/cloud_topics/level_one/frontend_reader/reader.cc:
```diff
@@ -0,0 +1,443 @@
+      _batches.empty(),
+      "Materialize batches called with batches already materialized");
+
+    if (_state == state::end_of_stream) {
```
from what i can tell can you assert here that the state is ready? for example, this is only called following a non-throwing execution of fetch_metadata which transitions the reader into the ready state?

EDIT: i see below you added this assert. but doesn't that mean you can remove this conditional?

wdberkeley replied:
`fetch_metadata` can transition to end-of-stream, so we can enter this function in end of stream state (kind of an artifact of the double switch instead of a loop).

dotnwat commented on src/v/cloud_topics/level_one/frontend_reader/reader.cc:
```diff
@@ -0,0 +1,443 @@
+        std::rethrow_exception(ex);
+    }
+
+    co_await reader->close();
```
can this fail? if so, what is the best way to handle it? ignore it, log it, mark state end of stream? closing readers after you're done with them is always a weird thing if they can throw. often just ignoring it can be reasonable?

wdberkeley replied:
It closes the output stream, which could throw 🤷 . I think the best thing to do is log and ignore. I will add that logic.

dotnwat commented on src/v/cloud_topics/level_one/frontend_reader/reader.cc:
```diff
@@ -0,0 +1,443 @@
+    // Second switch: enforces that the reader has materialized batches
+    // or reached end-of-stream.
+    switch (_state) {
+    case state::empty:
+    case state::ready:
+        _state = state::end_of_stream;
```
AFAICT you could vassert(false) that these states never occur, because the empty/ready state handlers in the switch above throw, except in the cases where the state eventually is transitioned into `materialized`...?

wdberkeley replied:
True, it's worth upgrading to vassert. The state should be either end-of-stream or materialized after the first switch.

dotnwat commented on src/v/cloud_topics/level_one/frontend_reader/reader.cc:
```diff
@@ -0,0 +1,443 @@
+
+    _config.start_offset
+      = dest->empty()
+          ? kafka::next_offset(_config.start_offset)
```
```
>     dest->swap(_batches);
>    _config.start_offset
>      = dest->empty()
>          ? kafka::next_offset(_config.start_offset)
```
something about this makes feels non-intuitive can you leave a comment explaining? for example, if `dest` is empty--ie we aren't consuming any batches--why is an offset incrementing? if it because re-requesting the same thing from the metastore would just return another empty set, then we should probably fix that in the metastore (not in this PR, but please do explain the situation as a comment in any case).

> _config.start_offset

should we keep the a class member `_next_offset` to track the current read pointer? using the config to store reader state feels a bit awkward because at any point in time it may not represent the configuration of the original read request.

wdberkeley replied:
Comment added ✅ 

You're right about tracking the read pointer outside of the config. It's a pattern in other readers but it's weird.

dotnwat replied:
> It's a pattern in other readers but it's weird.

i'm not sure i'd call it a pattern so much as it's just a random thing someone did :P

dotnwat reviewed (commented):
yeh i think this is very close. it looks great! i left some minor feedback, nothing major.

either now, or after the next round of feedback I do think we should squash this PR. there are currently 15 commits, and I think there should probably be 2 or 3:

1. add reader
2. add reader test (or add in (1))
3. whatever change is made to `fake_io.cc`

rockwotj commented on src/v/cloud_topics/level_one/frontend_reader/reader.cc:
```diff
@@ -0,0 +1,443 @@
+              _ntp,
+              _tidp);
+            _state = state::end_of_stream;
+            co_return;
```
this should fail loudly outright or can this happen when there has been no reconciliation yet @andrwng?

wdberkeley replied:
As we're imagining using this combined with an L0 reader, it shouldn't happen because the LRO should be at or before the end of the offset range spanned by L1 recorded in the metastore, and the higher level reader combining the L1 and L0 readers should provide a max offset to the L1 reader that will not trigger this case unless something is seriously wrong. But if one is using the L1 reader on its own it's expected one could hit this.

wdberkeley commented:
Force push to rebase on dev (squash incoming)

wdberkeley committed:
ct/l1: Add tests for L1 reader

Comprehensive test suite for the L1 reader covering:
- Empty reads when no data exists
- Reading from single and multiple L1 objects
- Multiple partitions in shared objects
- Offset range filtering
- Max bytes and strict_max_bytes behavior
- Out of range offset handling
- Error cases (missing objects)
- Empty offset ranges from compaction
- Sparse offset ranges with gaps

wdberkeley commented:
Force push for squash.

dotnwat commented on src/v/cloud_topics/level_one/frontend_reader/reader.cc:
```diff
@@ -0,0 +1,464 @@
+
+            // If we make it past all that, emit the batch.
+            _batches.push_back(std::move(batch));
+        } else if (std::holds_alternative<model::topic_id_partition>(result)) {
+            // Partition marker. Done with this ntp's partition.
+            break;
```
@rockwotj hmm, seems like this leaks details about the format/structure of the L1 object? is it intentional, or just a matter of having not yet built a better l1 object reader?

rockwotj replied:
yeah it leaks the details. We could update the API to only support reading a "segment" (data for a single NTP chunk) at a time, which would remove this boundary.

dotnwat replied:
got it. not a big deal just curious

rockwotj replied:
I'll think about a better API, but it's a bit hard todo and keep the reading decoupled from the mechanical details of how IO is fetched.

In practice we're always reading a single chunk so this should never actually happen.

dotnwat reviewed (approved):
i think all my feedback was addressed. thanks!

it can be often be pretty helpful on the reviewer side to coordinate on rebases that move the base commit because a common flow is to look at the diff between force pushes. `--keep-base` option to `rebase` does that. ofc doesn't work for fixing conflicts.
